### ISSUE TITLE
NPE on request.cancel()

### ISSUE THREAD

**Issue by sbernard31** [eclipse/californium#226]
I faced some NPE, when I cancel a request.
```
java.lang.NullPointerException: token bytes must not be null
	at org.eclipse.californium.core.network.Exchange$KeyToken.<init>(Exchange.java:637)
	at org.eclipse.californium.core.network.Exchange$KeyToken.fromOutboundMessage(Exchange.java:670)
	at org.eclipse.californium.core.network.UdpMatcher$ExchangeObserverImpl.completed(UdpMatcher.java:315)
	at org.eclipse.californium.core.network.Exchange.setComplete(Exchange.java:435)
	at org.eclipse.californium.core.network.stack.ExchangeCleanupLayer$CancelledMessageObserver.onCancel(ExchangeCleanupLayer.java:45)
	at org.eclipse.californium.core.coap.Message.setCanceled(Message.java:579)
	at org.eclipse.californium.core.coap.Request.setCanceled(Request.java:616)
	at org.eclipse.californium.core.coap.Message.cancel(Message.java:658)
```
I'm not sure but I suppose this could happen when request is canceled before token was assigned.

See the stack
```
 * +--------------------------+
 * | {@link MessageDeliverer}         |
 * +--------------A-----------+
 *                A
 *              * A
 * +------------+-A-----------+
 * |       CoAPEndpoint       |
 * |            v A           |
 * |            v A           |
 * | +----------v-+---------+ |
 * | | Stack Top            | |
 * | +----------------------+ |
 * | | {@link ExchangeCleanupLayer} | |
 * | +----------------------+ |
 * | | {@link ObserveLayer}         | |
 * | +----------------------+ |
 * | | {@link BlockwiseLayer}       | |
 * | +----------------------+ |
 * | | {@link ReliabilityLayer}     | |
 * | +----------------------+ |
 * | | Stack Bottom         | |
 * | +----------+-A---------+ |
 * |            v A           |
 * |          Matcher         |
 * |            v A           |
 * |        Interceptor       |
 * |            v A           |
 * +------------v-A-----------+
 *              v A 
 *              v A 
 * +------------v-+-----------+
 * | {@link Connector}                |
 * +--------------------------+
```
The `ExchangeCleanupLayer` add listener on `cancel` (which will call `exchange.complete`).
The `matcher` add listener on `exchange.complete` then attribute a token to the request.

More than the NPE, I suspect there is some scenario where exchangeStore will not be cleaned and token not release.

**Comment by sophokles73** [comment]
> More than the NPE, I suspect there is some scenario where exchangeStore will not be cleaned and token not release.

Why do you think so? Is there a particular scenario at hand in which this would happen?

**Comment by boaks** [comment]
Dig into #232 (branch 2.0.x).
Search for `MessageExchangeStore.releaseToken(KeyToken)`.
You find:

- BaseMatcher.cancelObserve(byte[]) 
  doesn't work for client observe exchanges as they are not stored in the `exchangesByToken` 

- matchNotifyResponse(ExchangeObserver, Response, CorrelationContext)
      new MessageObserverAdapter() {...}
   I think, only `onResponse` may be called.

- registerObserve(Request)
      new MessageObserverAdapter() {...}
   I think, this only work during the observe request.

- TcpMatcher / ExchangeObserverImpl.completed(Exchange) and
  UdpMatcher/ ExchangeObserverImpl.completed(Exchange)
  Both are only for `!exchange.getCurrentRequest().getOptions().hasObserve()`

So currently, I think, at least here is a leak.

**Comment by sophokles73** [comment]
Rebased on 2.0.x to include @boaks changes from #231 regarding *safe removal* from `MessageExchangeStore`.

**Comment by sophokles73** [comment]
NPE is fixed on 2.0.x in 79084c6493d426c15994f0bdf8ea77643cc5e9c6

Please open separate issues for suspected memory (token) leaks if feasible.
